socks5proxy/src/socks5client.cpp

```diff
@@ -0,0 +1,337 @@
+            sizeof(ClientConnectionAddressIpv4Packet)) {
+          m_inSocket->close();
+          return;
+        }
+
+        struct sockaddr_in sa;
+        sa.sin_family = AF_INET;
+        sa.sin_port = ntohs(packet.port);
+        memcpy(&(sa.sin_addr), (void*)&packet.ip_dst, 4);
+
+        m_outSocket = new QTcpSocket(this);
+        m_outSocket->connectToHost(QHostAddress((struct sockaddr*)&sa),
+                                   sa.sin_port);
+        configureOutSocket();
+      }
+
+      else if (m_addressType == 0x03 /* Domain name */) {
+        if (m_inSocket->bytesAvailable() == 0) return;
+
+        uint8_t length;
+        if (m_inSocket->read((char*)&length, 1) != 1) {
+          m_inSocket->close();
+          return;
+        }
+
+        char buffer[INT8_MAX];
+        if (m_inSocket->read(buffer, length) != length) {
+          m_inSocket->close();
+          return;
+        }
+
+        uint16_t port;
+        if (m_inSocket->read((char*)&port, sizeof(port)) != sizeof(port)) {
+          m_inSocket->close();
+          return;
+        }
+
+        m_outSocket = new QTcpSocket(this);
+        m_outSocket->connectToHost(QByteArray(buffer, length), ntohs(port));
+        configureOutSocket();
+      }
+
+      else if (m_addressType == 0x04 /* Ipv6 */) {
+        if (m_inSocket->bytesAvailable() <
+            (qint64)sizeof(ClientConnectionAddressIpv6Packet)) {
+          return;
+        }
+
+        ClientConnectionAddressIpv6Packet packet;
+        if (m_inSocket->read((char*)&packet,
+                             sizeof(ClientConnectionAddressIpv6Packet)) !=
+            sizeof(ClientConnectionAddressIpv6Packet)) {
+          m_inSocket->close();
+          return;
+        }
+
+        struct sockaddr_in6 sa;
+        sa.sin6_family = AF_INET6;
+        sa.sin6_port = ntohs(packet.port);
+        memcpy(&(sa.sin6_addr), (void*)&packet.ip_dst, 16);
+
+        m_outSocket = new QTcpSocket(this);
+        m_outSocket->connectToHost(QHostAddress((struct sockaddr*)&sa),
+                                   sa.sin6_port);
+        configureOutSocket();
+      }
+
+      else {
+        ServerResponsePacket packet(createServerResponsePacket(
+            0x08 /* Address type not supported */, 0x00));
+        m_inSocket->write((char*)&packet, sizeof(ServerResponsePacket));
+        m_inSocket->close();
+        return;
+      }
+    }
+
+    break;
+
+    case Proxy: {
+      qint64 bytes = proxy(m_inSocket, m_outSocket);
+      if (bytes) {
+        emit m_parent->dataSentReceived(bytes, 0);
+      }
+    } break;
+
+    default:
+      Q_ASSERT(false);
+      break;
+  }
+}
+
+qint64 Socks5Client::proxy(QTcpSocket* a, QTcpSocket* b) {
+  Q_ASSERT(a && b);
+
+  qint64 bytes = 0;
+  char buffer[4096];
+  while (a->bytesAvailable()) {
+    qint64 val =
+        a->read(buffer, qMin(a->bytesAvailable(), (qint64)sizeof(buffer)));
+    if (val <= 0 || b->write(buffer, val) != val) {
```

Comment:
This does leave open the possibility that too much incoming data can overwhelm the outbound socket when connecting to a slow (or unreachable?) host. In such a situation this could result in unexpected connection drops when the more correct behaviour is to throttle the reception of data on the incoming socket.